### ISSUE TITLE
Would there be any way of fixing the value of specific weights in nn.AnalogLinear?

### ISSUE THREAD

**Issue by nkyungmi** [IBM/aihwkit#150]
Hi! I'm trying to modify the values of specific weights in `nn.AnalogLinear` layer.
I'm currently using MLP (784-256-128-10) and each layer is made up of `nn.AnalogLinear`
I know how to access each weight values and initialize, but what I want to do is not only to modify the values but to fix the values for the entire training process, without effecting the gradients. 
Would there be any way of doing this using AnalogLinear? 

(Sorry for not properly adding the label 'question', that I'm so new to this github system..!)

**Comment by maljoras** [comment]
Thank you very much for your question, @nkyungmi!  If I understand your question correctly, then you do not want to train the analog weights at all, but maybe some other (digital) parts of the DNN? 

In principle, one can just set the learning rate of a particular analog layer to zero, however, when using the `AnalogSGD`, the overall learning rate of the optimizer will be used instead.     

We do not support different learning rates for analog and digital components in our `AnalogSGD` for now (we might add this in near future), but one quick (and dirty) way to achieve a similar thing is to *not* use the `AnalogSGD` but instead use one of pytorch's native optimizers (e.g. `torch.optim.SGD`) and also omit the parameter regrouping (ie delete this line https://github.com/IBM/aihwkit/blob/master/examples/01_simple_layer.py#L47).  This should then fix the analog weights, as the optimizer is not aware of the analog features, but should still train other (digital) parts of the DNN. Note, however, that in this case any decay, diffusion or drift process is omitted as well, since that is usually handled in the `AnalogSGD`.  Other non-linearities and noises of the forward and backward pass will still be present though. 

In case that you do not want to train the network at all, but only compute the backward pass, then you could just omit the optimizer step altogether (i.e. delete the `opt.step()`).

**Comment by maljoras** [comment]
In case you want to only train a specific layer, e.g. let's say you have   
```python
model = AnalogSequential(
        AnalogLinear(input_size, hidden_sizes[0], True,
                     rpu_config=SingleRPUConfig(device=ConstantStepDevice())),
        nn.Sigmoid(),
        AnalogLinear(hidden_sizes[0], hidden_sizes[1], True,
                     rpu_config=SingleRPUConfig(device=ConstantStepDevice())),
        nn.Sigmoid(),
        AnalogLinear(hidden_sizes[1], output_size, True,
                     rpu_config=SingleRPUConfig(device=ConstantStepDevice())),
        nn.LogSoftmax(dim=1)
    )
```
Then there is currently not a good way to, e.g. fix the first layer and only train the others. It is a good idea, and we will try to support this in future. 

If you cannot wait and want to take things in your own hand, and you feel comfortable enough to edit the source code (and assuming you have updated to the last version of the master branch) you could just replace this line (https://github.com/IBM/aihwkit/blob/master/src/aihwkit/optim/analog_sgd.py#L48):
```python
module.analog_tile.set_learning_rate(self.defaults['lr'])
```
with:
```python 
if module.analog_tile.get_learning_rate() > 0:
    module.analog_tile.set_learning_rate(self.defaults['lr'])
````
Then you could turn off the learning of a specific layer by setting the LR to zero manually before training. For example, in case of the above model definition,
```python
 model[0].analog_tile.set_learning_rate(0)
```
would then set the LR of the first layer to zero throughout the training and thus fix its weights.

**Comment by maljoras** [comment]
@nkyungmi I have to correct my answers above, it is actually already possible to specify a different learning rate for each layer, or turn off learning for all analog layer!

In pytorch, all layers have their own learning rate in fact. So you can indeed use our `AnalogSGD` and specify learning rates for each layer. For instance if you want to set all analog learning rates to zero, you can do (after you regroup the parameter as in the examples):
```python
for group in opt.param_groups:
    if 'analog_tile' in group:
        group['lr'] = 0.0
```
that would turn off all the analog tiles' learning. Similarly, if you only want to turn off the first analog layers learning, you could define:
```python
def set_lr_analog_layer(analog_opt, lr, idx):
    """ Set the learning rate of the n-th analog tile defined in the model.
    Params: 
        analog_opt: analog optimizer (already re-grouped) 
        lr:  updated learning rate
        idx: analog layer index (only counting analog layers)
    """  
    n_layer=0
    for group in analog_opt.param_groups:
        if 'analog_tile' not in group:
            continue 
        if n_layer == idx:
            group['lr'] = lr
            break
        n_layer += 1
            
```
and then
```python
set_lr_analog_layer(opt, lr=0.0, idx=0)
``` 
after the definition of the optimizer and before the training. That should set only the LR of the first analog layer to 0, and thus fix the weight values.

**Comment by nkyungmi** [comment]
Thank you for your very detailed answer!! 
Yes, I was trying to train the network outside this analog tiles then put weight values back to the layer. 
I now got that I can stop training for some layers in the network by using the method you mentioned above. 
I'll try that. 
Thank you again :)!!
